pkg/components/rook/component_conversion_test.go

```diff
@@ -0,0 +1,227 @@
+					"rook-ceph/templates/service-monitor.yaml",
+				}
+
+				for _, ec := range expectedConfig {
+					if _, ok := m[ec]; !ok {
+						t.Fatalf("Expected %q to be generated but it is not available", ec)
+					}
+				}
+			},
+		},
+		{
+			"OperatorTolerations", func(t *testing.T, m map[string]string) {
+				expected := []corev1.Toleration{
+					{Key: "toleration_key1", Operator: "Equal", Value: "toleration_value1", Effect: "NoSchedule"},
+					{Key: "toleration_key2", Operator: "Equal", Value: "toleration_value2", Effect: "NoSchedule"},
+				}
+				got := deploy.Spec.Template.Spec.Tolerations
+
+				if !reflect.DeepEqual(expected, got) {
+					t.Fatalf("Expected: %v, got: %v", expected, got)
+				}
+			},
+		},
+		{
+			"EnvVars", func(t *testing.T, m map[string]string) {
+				// nolint:lll
+				subTCs := map[string]string{
+					"AGENT_TOLERATION_KEY":          "agent_toleration_key",
+					"AGENT_TOLERATION":              "NoSchedule",
+					"AGENT_TOLERATIONS":             "- effect: NoSchedule\n  key: toleration_key1\n  operator: Equal\n  value: toleration_value1\n- effect: NoSchedule\n  key: toleration_key2\n  operator: Equal\n  value: toleration_value2",
+					"DISCOVER_TOLERATION_KEY":       "discover_toleration_key",
+					"DISCOVER_TOLERATION":           "NoSchedule",
+					"DISCOVER_TOLERATIONS":          "- effect: NoSchedule\n  key: toleration_key1\n  operator: Equal\n  value: toleration_value1\n- effect: NoSchedule\n  key: toleration_key2\n  operator: Equal\n  value: toleration_value2",
+					"CSI_PROVISIONER_TOLERATIONS":   "- effect: NoSchedule\n  key: toleration_key1\n  operator: Equal\n  value: toleration_value1\n- effect: NoSchedule\n  key: toleration_key2\n  operator: Equal\n  value: toleration_value2",
+					"CSI_PLUGIN_TOLERATIONS":        "- effect: NoSchedule\n  key: other_node_key1\n  operator: Equal\n  value: other_node_value1",
+					"AGENT_NODE_AFFINITY":           "storage_node_key1=storage_node_value1; storage_node_key2=storage_node_value2;",
+					"DISCOVER_AGENT_NODE_AFFINITY":  "storage_node_key1=storage_node_value1; storage_node_key2=storage_node_value2;",
+					"CSI_PROVISIONER_NODE_AFFINITY": "storage_node_key1=storage_node_value1; storage_node_key2=storage_node_value2;",
+					"CSI_PLUGIN_NODE_AFFINITY":      "node_key1=node_value1; node_key2=node_value2;",
+				}
+
+				envs := deploy.Spec.Template.Spec.Containers[0].Env
+				envVars := make(map[string]string, len(envs))
+				for _, en := range envs {
+					envVars[en.Name] = en.Value
+				}
+
+				for k, v := range subTCs {
+					k, v := k, v
+					t.Run(k, func(t *testing.T) {
+						gotVal, ok := envVars[k]
+						if !ok {
+							t.Fatalf("Expected: %q env", k)
+						}
+
+						if gotVal != v {
+							t.Fatalf("Expected env var %q:%q, got %q:%q", k, v, k, gotVal)
+						}
+					})
+				}
+			},
+		},
+	}
+
+	for _, tc := range tcs {
+		tc := tc
+		t.Run(tc.Name, func(t *testing.T) {
+			t.Parallel()
+			tc.Test(t, m)
```

Comment:
After seeing @johananl comments and looking at this code, This indeed doesn't seem to make sense to define those tests in the array?

You could just have a helper function returning `deploy` and `m` and have separate test cases with proper names.

That would be simpler case if there is no logic shared between the tests at the end.

Reply:
What's wrong in defining them in an array? Which test case has "improper" names?

Reply:
I think it's just too complex for no reason and inconsistent with other tests we have written. We could write all tests we have this way.

Reply:
IMHO this way of writing code is more readable because you can look at the array and understand what is going on. All the boilerplate can be ignored for a reader.

Reply:
I think table-driven tests (tests cases in a slice) are a very good practice in Go. It's also recommended [officially](https://github.com/golang/go/wiki/TableDrivenTests) and [unofficially](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests). Besides being easy to read and maintain, defining all the test-case-specific data in a declarative data structure forces you to think about your tests cases in a more structured - and generic - way.

Reply:
Yes, I agree table-driven tests are good, but if they share common test logic, which is not the case here, as we only share preparation logic and we define testing function in a table... All examples you linked @johananl actally share the test logic.

Reply:
My point was that perhaps we *can* structure the tests so that they have a shared logic. I didn't try to refactor the tests myself and don't want to halt the merging for this, just something for @surajssd to think about in case there is an obvious way to improve how we're doing things here.

Reply:
Changed the code to be in different functions which resue the boilterplate.